### PR TITLE
Restore AES Key File on database restoration

### PR DESCRIPTION
**Pull request by thewhitewizard**

### CODE REVIEW

**Review by mcornaton — APPROVED** [review]

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 1]
```diff
@@ -40,6 +41,7 @@
 @Service
 public class AdminService {
 
+    private static final String AES_KEY_LOG_DESCRIPTION = "AES Key";
```
Comment (jbern0rd) [inline src/main/java/com/iexec/sms/admin/AdminService.java:44]: Really minor, I tend to prefer gathering constants together for clarity.
Would it be possible to have this definition next to `AES_KEY_FILENAME_EXTENSION` definition ?

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 2]
```diff
@@ -284,16 +337,17 @@ private void checkDestinationFileNotExists(Path destination, AdminOperationError
...
+     * @param standardCopyOption Copy option
      * @throws IOException If the copy fails
      */
-    private void processCopyFile(Path source, Path destination, String fileDescription) throws IOException {
+    private void processCopyFile(Path source, Path destination, String fileDescription, StandardCopyOption standardCopyOption) throws IOException {
         final long size = source.toFile().length();
         log.info("{} copy process start [source:{}, destination :{}]", fileDescription, source, destination);
         final long start = System.currentTimeMillis();
-        Files.copy(source, destination, StandardCopyOption.COPY_ATTRIBUTES);
+        Files.copy(source, destination, standardCopyOption);
```
Comment (jbern0rd) [inline src/main/java/com/iexec/sms/admin/AdminService.java:350]: Don't we want to preserve the vararg aspect of `Files.copy` here and have both `REPLACE_EXISTING` and `COPY_ATTRIBUTES` when copying the key ?
See https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Files.html#copy(java.nio.file.Path,java.nio.file.Path,java.nio.file.CopyOption...)

File: src/main/java/com/iexec/sms/admin/OutOfServiceRequestFilter.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+/**
+ * This filter can be used to disable incoming requests.
+ */
+public class OutOfServiceRequestFilter extends GenericFilterBean {
+
+    public static final String MAINTENANCE_ERROR = "The server is temporarily unable to service your request due to maintenance downtime";
```
Comment (jbern0rd) [inline src/main/java/com/iexec/sms/admin/OutOfServiceRequestFilter.java:32]:
```suggestion
    public static final String MAINTENANCE_ERROR = "The server is temporarily unable to serve your request due to maintenance downtime";
```

File: src/main/java/com/iexec/sms/config/OufOfServiceFilterConfig.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+import com.iexec.sms.admin.AdminService;
+import com.iexec.sms.admin.OutOfServiceRequestFilter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression;
+import org.springframework.boot.web.servlet.FilterRegistrationBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ConditionalOnExpression("'${admin.outofservice.enabled}'=='true'")
```
Comment (jbern0rd) [inline src/main/java/com/iexec/sms/config/OufOfServiceFilterConfig.java:26]: Should we go for something like `admin.out-of-service.enabled` in `application.yml` configuration file for more readability ?

Comment (jbern0rd) [reply src/main/java/com/iexec/sms/config/OufOfServiceFilterConfig.java:26]: See https://docs.spring.io/spring-boot/docs/2.7.17/reference/html/features.html#features.external-config.typesafe-configuration-properties.relaxed-binding.environment-variables

File: src/main/java/com/iexec/sms/config/OufOfServiceFilterConfig.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+public class OufOfServiceFilterConfig {
```
Comment (jbern0rd) [inline src/main/java/com/iexec/sms/config/OufOfServiceFilterConfig.java:27]: Could we add some documentation to list the patterns that will still be served ?
`/admin`, `/version`, `/metrics`, `/swagger-ui/index.html`, `/v3/api-docs`

File: src/main/java/com/iexec/sms/encryption/EncryptionService.java [hunk 6]
```diff
@@ -120,4 +130,23 @@ boolean checkOrFixReadOnlyPermissions(String aesKeyPath) {
...
         return true;
     }
+
+    /**
+     * Enables the file to be written
+     * Use for restoration process
+     *
+     * @return {@code true} if the switch was successful, {@code false} if any error occurs.
+     */
+    public boolean setWritePermissions() {
```
Comment (jbern0rd) [inline src/main/java/com/iexec/sms/encryption/EncryptionService.java:140]: Wouldn't we prefer to return `File.canWrite()` in this method ?
What is more important to us in this specific situation, the goal or the journey ?

Comment (jbern0rd) [reply src/main/java/com/iexec/sms/encryption/EncryptionService.java:140]: I mean this would return `false` if the file was already writable

Comment (jbern0rd) [reply src/main/java/com/iexec/sms/encryption/EncryptionService.java:140]: Or should we add a protection as done in `checkOrFixReadOnlyPermissions` ?

File: src/main/java/com/iexec/sms/admin/AdminService.java (hunk 1 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:44]: done locally

File: src/main/java/com/iexec/sms/admin/OutOfServiceRequestFilter.java (hunk 3 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/OutOfServiceRequestFilter.java:32]: done locally

File: src/main/java/com/iexec/sms/config/OufOfServiceFilterConfig.java (hunk 5 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/config/OufOfServiceFilterConfig.java:27]: done locally

File: src/main/java/com/iexec/sms/admin/AdminService.java (hunk 2 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:350]: it could be a good idea

**Review by jbern0rd — APPROVED** [review]
Thanks

**Pull request merged**